File: code/python/13.Relay/Relay.py
import RPi.GPIO as GPIO
import time

relayPin = 11    # define the relayPin
buttonPin = 12    # define the buttonPin

def setup():
    print ('lnitialization')
    GPIO.setmode(GPIO.BOARD)       # Numbers GPIOs by physical location
    GPIO.setup(relayPin, GPIO.OUT)   # Set relayPin's mode is output
    GPIO.setup(buttonPin, GPIO.IN)   # Set buttonpin mode input

def loop():
    i=0
    g=0
    while True:
        reading = GPIO.input(buttonPin)
        if reading == GPIO.HIGH & g==0:
            time.sleep(0.2)
            reading = GPIO.input(buttonPin)
            if reading == GPIO.HIGH:
                i=i+1
                i=i%2
                g=1
                if i==0:
                    GPIO.output(relayPin,GPIO.LOW)
                    print("relayPin.....off")
                if i==1:
                    GPIO.output(relayPin,GPIO.HIGH)
                    print("relayPin.....on")
        else:
            g=0
    
def destroy():
    GPIO.output(relayPin, GPIO.LOW)     # relay off
    GPIO.cleanup()                     # Release resource

if __name__ == '__main__':     # Program start from here
    setup()
    try:
        loop()
    except KeyboardInterrupt:  
        destroy()
        

File: code/C/13.Relay/Relay.c
#include <wiringPi.h>
#include <stdio.h>

#define relayPin    0  	//define the relayPin
#define buttonPin 1		//define the buttonPin
int main(void)
{
	if(wiringPiSetup() == -1){ //when initialize wiring fairelay,print messageto screen
		printf("fairelay !");
		return 1; 
	}
	printf("starting...\n");
	pinMode(relayPin, OUTPUT); 
	pinMode(buttonPin, INPUT);
	pullUpDnControl(buttonPin, PUD_UP);  //pull up to high level
	int i=0;
	int g=0;
	while(1){
		if(digitalRead(buttonPin)==HIGH && g==0)
		{
	            delay(20);
	            if(digitalRead(buttonPin)==HIGH)
		    {
	                i=i+1;
	                i=i%2;
	                g=1;
	                if(i==0)
			{
	                    digitalWrite(relayPin,LOW);
	                    printf("relayPin.....off\n");
			}
	                if(i==1)
			{
	                    digitalWrite(relayPin,HIGH);
	                    printf("relayPin.....on\n");
			}
		     }
		}
	        else if(digitalRead(buttonPin)==LOW)
		{
            		g=0;
		}
	}

	return 0;
}
